File: Year-2019/Mar/2019-03-14-189/watercow/189.c
// 看到一个厉害的思路
/* *
* 二次旋转
* Original List                   : 1 2 3 4 5 6 7
* After reversing all numbers     : 7 6 5 4 3 2 1
* After reversing first k numbers : 5 6 7 4 3 2 1
* After revering last n-k numbers : 5 6 7 1 2 3 4 --> Result
* 旋转函数是二分交换元素，然后头++，尾--
* 时间复杂度O(n)
* 空间复杂度O（1)
* */

class Solution {
public:
	void rotate(vector<int>& nums, int k) {
		
		//首先k要保证小于nums.size，进行一个模运算
		k = k % nums.size();

		reverse(nums, 0, nums.size() - 1);
		reverse(nums, 0, k - 1);
		reverse(nums, k, nums.size() - 1);
	}

	void reverse(vector<int>& nums, int start, int end) {
		while (start < end) {
			int temp = nums[start];
			nums[start++] = nums[end];
			nums[end--] = temp;
		}
	}
};

File: Year-2019/Mar/2019-03-14-189/lb/189.java
class Solution {
    /**
     * 学习一个内置函数
     * System.arraycopy(tmp,tmp.length-k,nums,0,k)
     * 把temp中序号为length-k起的k个数复制到num中序号为0起的k个数
     * @param nums
     * @param k
     */
    public void rotate(int[] nums, int k) {
        int length = nums.length;
        k %= length;
        int left = length-k;
        int[] temp = new int[left];
        //备份前length - k 个数
        for (int i = 0;i<left;i++){
            temp[i] = nums[i];
        }
        //旋转后面k个数
        for(int i = left;i<length;i++){
            nums[i-left] = nums[i];
        }
        //填入前k个数
        for (int i=k;i<length;i++){
            nums[i] = temp[i-k];
        }
//        k=k%nums.length;
//        int[] tmp=nums.clone();
//        System.arraycopy(tmp,tmp.length-k,nums,0,k);
//        System.arraycopy(tmp,0,nums,k,tmp.length-k);
    }
}
